src/transformers/models/gpt_neo/modeling_gpt_neo.py

```diff
@@ -371,115 +337,15 @@ def forward(
         return outputs  # a, present, (attentions)
 
 
-class GPTNeoLocalSelfAttention(nn.Module, GPTNeoAttentionMixin):
```

Comment:
Why delete all this?

Reply:
The way local attention was implemented caused very high VRAM use with certain, especially prime, sequence lengths as described in #11320:

![patched](https://user-images.githubusercontent.com/82650881/115746923-0ee0cc00-a395-11eb-9bcd-c289a38c77b5.png)

The green line is the original implementation's VRAM use in GB over sequence length. Padding inputs to multiples of the local attention window sizes mitigates this, but the code is still hard to follow. The way local attention is implemented in this PR follows PR #11630. Instead of the very complex folding approach, it is calculated like global attention with an adjusted mask, which masks out attention values beyond the local attention window. This is much simpler, easier to understand and does not seem to have any noticable negative performance impact. It also fixes the issue with high VRAM usage.

Reply:
I see, it's mentioned in the other PR! Think we can go along with this after having done some memory & inference benchmarks

Reply:
I'm currently running benchmarks of both implementations. Will report back once they finish.

Reply:
You can find the full results in these notebooks:

[localattention_4.6.1.ipynb](https://github.com/finetuneanon/misc/blob/main/transformers_benchmark/localattention_4.6.1.ipynb)
[localattention_gpt-j.ipynb](https://github.com/finetuneanon/misc/blob/main/transformers_benchmark/localattention_gpt-j.ipynb)

My implementation of local attention seems to be significantly faster. It uses slightly more memory for sequence lengths that are divisible by the window size, but exhibits no sudden jumps in VRAM usage like the current implementation.

4.6.1, half precision:
```
torch.float16 cuda:0
seq_len	max_len	runtime	memory
64	104	1.2820s	5.059GB
127	167	1.4140s	5.090GB
251	291	1.8570s	5.181GB
256	296	1.8761s	5.180GB
383	423	2.8156s	6.994GB
509	549	3.3879s	7.646GB
512	552	2.7486s	5.317GB
769	809	4.6178s	9.000GB
1021	1061	5.7816s	10.304GB
1024	1064	4.4912s	5.590GB
1280	1320	5.3699s	5.735GB
1283	1323	7.0542s	11.664GB
1536	1576	6.2600s	5.882GB
1543	1583	8.2814s	13.013GB
1792	1832	7.1836s	6.023GB
1801	1841	9.5767s	14.350GB
[OOM error]
```

gpt-j branch, half precision:
```
torch.float16 cuda:0
seq_len	max_len	runtime	memory
64	104	0.9781s	5.186GB
127	167	1.0396s	5.225GB
251	291	1.1590s	5.310GB
256	296	1.1628s	5.311GB
383	423	1.3393s	5.401GB
509	549	1.4969s	5.490GB
512	552	1.4975s	5.493GB
769	809	1.8752s	5.683GB
1021	1061	2.1957s	5.861GB
1024	1064	2.1975s	5.861GB
1280	1320	2.5928s	6.051GB
1283	1323	2.6369s	6.053GB
1536	1576	2.9966s	6.243GB
1543	1583	3.0621s	6.249GB
1792	1832	3.3844s	6.422GB
1801	1841	3.4636s	6.426GB
2039	2049	2.6127s	6.597GB
2048	2049	2.2140s	6.458GB
```